kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateCloudRepository.java

```diff
@@ -0,0 +1,151 @@
+
+    public static XStream initializeXStream() {
+        XStream xs = XStreamUtils.createTrustingXStream(new PureJavaReflectionProvider());
+        String[] voidDeny = {"void.class", "Void.class"};
+        xs.denyTypes(voidDeny);
+        xs.alias(CFG_MAP_DATA_KEY, KieServerState.class);
+        xs.alias("container", KieContainerResource.class);
+        xs.alias("config-item", KieServerConfigItem.class);
+
+        return xs;
+    }
+
+    public KieServerStateCloudRepository() {
+        xs = initializeXStream();
+    }
+
+    @Override
+    public synchronized void store(@NotNull String serverId, @NotNull KieServerState kieServerState) {
+
+        String kssServerId = null;
+
+        try {
+            kssServerId = kieServerState.getConfiguration()
+                                        .getConfigItem(KieServerConstants.KIE_SERVER_ID).getValue();
+        } catch (Exception e) {
+            logger.error("Failed to retrieve server id from  KieServerState", e);
+        }
+
+        if (kssServerId == null || serverId.length() == 0) {
+            throw new IllegalArgumentException("Invalid KieServerId: Can not be null or empty.");
+        }
+
+        if (!kssServerId.equals(serverId)) {
+            throw new IllegalArgumentException("Invalid KieServerId: Id does not match with KieServerState.");
+        }
+
+        try (KubernetesClient client = createKubernetesClient()) {
+
+            ConfigMap cm = client.configMaps().withName(serverId).get();
+
+            if (cm == null) {
+                throw new IllegalStateException("KieServerState ConfigMap must exist before save/update.");
+            }
+
+            String stateXML = xs.toXML(kieServerState);
+            ObjectMeta md = cm.getMetadata();
+
+            Map<String, String> ann = md.getAnnotations() == null ? new ConcurrentHashMap<>() : md.getAnnotations();
+            md.setAnnotations(ann);
+            ann.put(STATE_CHANGE_TIMESTAMP,
+                    ZonedDateTime.now().format(DateTimeFormatter.ISO_INSTANT));
+
+            if (isKieServerReady()) {
+                ann.put(ROLLOUT_REQUIRED, "true");
+            }
+
+            Map<String, String> cmData = new ConcurrentHashMap<>();
+            cmData.put(CFG_MAP_DATA_KEY, stateXML);
+            cm.setData(cmData);
+
+            client.configMaps().createOrReplace(cm);
+
+        } catch (IllegalStateException ise) {
+            throw ise;
+        } catch (Exception e) {
+            logger.error("Storing KieServerState failed.", e);
+        }
+
+    }
+
+    @Override
+    @NotNull
+    public KieServerState load(@NotNull String serverId) {
+        KieServerState kieServerState = null;
+
+        try (KubernetesClient client = createKubernetesClient()) {
+            Map<String, String> data = client.configMaps()
+                                             .withName(serverId)
+                                             .get()
+                                             .getData();
```

Comment:
This line produces NPE in case server id to be loaded doesn't exists. I think it would be better to avoid NPE by checking Config map, returning more specific exception (for example "KieServerId " + serverId + " not found. Please create a config map with configuration first.").

Reply:
Refactored code including suggested change.